### ISSUE TITLE
FYI: Feature Policies [headers]

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#464]
- article: https://scotthelme.co.uk/a-new-security-header-feature-policy/
- spec: https://wicg.github.io/feature-policy/
- github (features): https://github.com/WICG/feature-policy/blob/master/features.md#policy-controlled-features
- a bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1390801
- a uMatrix issue: `https://github.com/uBlockOrigin/uMatrix-issues/issues/30`

So I assume (once implemented in FF) we could manipulate headers to control these. However, IMO, for now, the feature list is rather boring - most of these can already be controlled on a site by site basis (geo, camera etc), or are disabled (vibrate etc - who the hell wants vibrate?). But some could be handy - eg being able to block "full-screen" but allow eg YouTube. Autoplay also looks handy

@gorhill If you build this into uM at some stage, would I wrong to assume you would just pick the important/useful ones and add to the three-dot menu?

**Comment by gorhill** [comment]
I haven't decided anything yet, I need to do more thinking and prototype to make decisions.

**Comment by KOLANICH** [comment]
I guess we don't need to manipulate this instead of hooking. You know, Android permissions were revokable using root and `pm` command long before 6, but revoking them caused apps crashing with null pointer exceptions. So faking permissions and other dangerous API is a better approach. So we may need this only as a defence in depth measure for the case the app managed to bypass our hooks.

**Comment by Thorin-Oakenpants** [comment]
per domain autoplay is coming too - https://old.reddit.com/r/firefox/comments/90jgd2/nightly_now_lets_you_allow_or_block_autoplay_per/

I guess I'll learn (read and try to understand) more about it later - but if someone wants to ELI5, go for it
- The header is set by who (the website, or just locally?)
- Where does the header go - is it coming from the site and modified or added out of thin air by the extension - and then read by FF internally?
- If the header is sent, this w/could raise entropy? Surely its just internal/local? Right guys (and gals)?
- I have more questions but meh, there's so little to excite me for now, but I'd like to understand the mechanism better
- TIA

**Comment by Atavic** [comment]
These headers are set by the site's server, as a response to the request of your FF.

**Comment by Thorin-Oakenpants** [comment]
^^ so... my FF "requested" them by talking to the server?.. so, in other words, more entropy if they desired to track it. Still not seeing anything really useful here (fullscreen and maybe WebRTC if it lands), but the list can grow .. guess we'll see

**Comment by Atavic** [comment]
Yes, HTML Headers are sent everytime you load an HTML Document. The server sends HTML Headers back to your requests - eventually - with these security features enabled. It's all a server-browser dialogue: if your browser understands these features, it will follow these security feats.

**Comment by Atavic** [comment]
A list of Feature Policy values: https://www.owasp.org/index.php/OWASP_Secure_Headers_Project#fp
